turbopack/crates/turbo-tasks/src/effect.rs

```diff
@@ -168,16 +171,23 @@ pub async fn apply_effects(source: impl CollectiblesSource) -> Result<()> {
     }
     let span = tracing::info_span!("apply effects", count = effects.len());
     async move {
-        effects
+        let effects = effects
             .into_iter()
             .map(async |effect| {
                 let Some(effect) = ResolvedVc::try_downcast_type::<EffectInstance>(effect) else {
                     panic!("Effect must only be implemented by EffectInstance");
                 };
                 effect.await?.apply().await
             })
-            .try_join()
-            .await?;
+            // TODO remove this collect(), but rust was not happy with it...
+            .collect::<Vec<_>>();
+        // Limit the concurrency of effects,
+        // run them all even if an error occurs,
+        // report the first error.
+        let mut results = futures::stream::iter(effects).buffered(APPLY_EFFECTS_CONCURRENCY_LIMIT);
+        while let Some(result) = results.next().await {
+            result?;
+        }
```

Comment:
I think https://docs.rs/futures/latest/futures/prelude/stream/trait.TryStreamExt.html#method.try_for_each_concurrent might be a better fit for this.

Reply:
There is a difference in error handling. `try_for_each_concurrent` will cancel after the first error and also report the first error by timing (race condition) instead of the first error in order. But that sounds acceptable in this use case.